tests: add streamlit tests for most home bar items

naterush opened the pull request:
# Description

Adds tests for home-bar items. This is supposed to be testing that all basic functionality works -- and doing so in a way that is robust and easy to maintain.

# Testing

See tests. We want them to pass the following criteria:


1. All tests can run in parallel with no issues (two in parallel) currently.
2. Across 10 test runs (which we add with the following), we pass with *no flakeyness*
    
    ```
    strategy:
          matrix:
            run_count: [1, 2, ..., 10]
    ```
    
3. The tests themselves should run in <1 minute. This includes only the *testing* time -- this is true currently.

# Documentation

No.

aarondr77 commented on tests/streamlit_ui_tests/streamlit.spec.ts:
```diff
@@ -1,9 +1,302 @@
+    
+    await checkOpenTaskpane(mito, 'Conditional Formatting');
+    
+    // This test is currently broken, for some reason this button won't click
+    await mito.getByRole('button', { name: 'Add Conditional Formatting Rule' }).click();
```
Given the tests are passing, I suppose this comment is out of date?

naterush replied:
Tbh, skipping this test currently. Struggled to get it working, so will circle back to it when we've got a bit more experience :)

aarondr77 commented on tests/streamlit_ui_tests/streamlit.spec.ts:
```diff
@@ -1,9 +1,302 @@
+  test.skip('Color Dataframe', async ({ page }) => {
+    const mito = await getMitoFrameWithTestCSV(page);
+
+    await clickButtonAndAwaitResponse(page, mito, { name: 'Format', exact: true })
```
How come you sometimes use `clickButtonAndAwaitResponse` and other times just use `getByRole.....click()`

naterush replied:
Added to README.

aarondr77 commented on tests/streamlit_ui_tests/streamlit.spec.ts:
```diff
@@ -1,9 +1,302 @@
+    await mito.locator('.mito-dropdown-item').first().click();
+    await expect(mito.locator('.plotly-graph-div').first()).toBeVisible();
+  });
+
```
There are some more tests for the home tab that might be reasonable to add:
1. If there is no data in the tool, then you shouldn't be able to click pivot, merge, etc. 
2. Test that the AI feature opens when the button is clicked
3. Check that the AI button is not displayed on enterprise 
4. Check that the increase/decrease decimal buttons work 
5. Check the formatting select works

naterush replied:
- [ ] If there is no data in the tool, then you shouldn't be able to click pivot, merge, etc.
- [ ] Test that the AI feature opens when the button is clicked
- [ ] Check that the AI button is not displayed on enterprise
- [ ] Check that the increase/decrease decimal buttons work
- [ ] Check the formatting select works

naterush replied:
These other two were hard to write, so skipping for now. We'll keep coming back to them!

aarondr77 commented on tests/streamlit_ui_tests/streamlit.spec.ts:
```diff
@@ -1,9 +1,302 @@
+    await awaitResponse(page);
+
+    // Expect there to be 4 column headers
+    await expect(mito.locator('.endo-column-header-container')).toHaveCount(4);
```
Want to test that its in the correct location and selected also?

aarondr77 commented on tests/streamlit_ui_tests/streamlit.spec.ts:
```diff
@@ -1,9 +1,302 @@
+    await mito.getByText('Or', { exact: true }).click();
+    await mito.getByText('>').nth(1).click();
+    await mito.getByText('<').click();
+    await mito.locator('div').filter({ hasText: /^Or<$/ }).getByRole('textbox').fill('10');
```
Want to test that the `result` text is updated correctly? The text that says `Removed an additional XXX rows`

aarondr77 commented on tests/streamlit_ui_tests/streamlit.spec.ts:
```diff
@@ -1,9 +1,302 @@
+    await expect(mito.getByText('100.0%')).toBeVisible();
+    await clickButtonAndAwaitResponse(page, mito, 'Decrease the number of decimal places that are displayed in the selected number columns.')
+    await expect(mito.getByText('100%')).toBeVisible();
+    await mito.getByText('100%').click();
```
How come this ends by clicking on the text? I don't think we do this in the other tests

aarondr77 commented on tests/streamlit_ui_tests/streamlit.spec.ts:
```diff
@@ -1,9 +1,302 @@
+    // background-color: rgba(179, 41, 41, 0.4)
+    await expect(mito.locator('.mito-grid-cell').filter({ hasText: /^2$/ }).first()).not.toHaveAttribute('style', /background-color: rgba\(179, 41, 41, 0.4\)/);
+  
+    await mito.getByRole('button', { name: 'Import Files' }).click();
+    await mito.getByText('test.csv').click();
+    await mito.getByText('test.csv', { exact: true }).click();
+    await mito.locator('div').filter({ hasText: /^test\.csv$/ }).nth(1).click();
+    await mito.getByRole('button', { name: 'Import test.csv' }).click();
+    await mito.getByRole('button', { name: 'Import test.csv' }).click();
+    await mito.locator('#mito-center-content-container line').nth(1).click();
+    await mito.getByText('Column3').click();
+    await mito.locator('[id="mito-toolbar-button-add\\ column\\ to\\ the\\ right"]').getByRole('button', { name: 'Insert' }).click();
+    await mito.locator('#cell-editor-input').fill('=123');
+    await mito.locator('#cell-editor-input').press('Enter');
```
I don't think this code belongs in the `Conditional Formatting` test

aarondr77 reviewed (commented):
Nice! This is a great start to testing. Excited to keep developing these. Even if we don't get to the point where tests capture bugs like this (https://github.com/mito-ds/mito/issues/1008) for a long time. If we start to speed up the release process, then we'll have more time to fix bugs like that that we know of.

A few comments below about other tests I think we should create, etc. 

Also, can you:
1.  clean out the `tests` folder. There's a lot of old Jupyter stuff around that we probably don't want anymore
2. add a readme to the streamlit folder that has instructions on things like running these tests locally, writing tests, etc.

naterush commented:
1. Many of the JupyterLab tests should still be around, but I removed those that duplicated functionality.
2. Updated the README to have instructions on basic test writing.
